src/Microsoft.AspNet.Hosting/Internal/RequestServicesContainerMiddleware.cs

```diff
@@ -21,9 +23,37 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceProvider
 
         public async Task Invoke(HttpContext httpContext)
         {
-            using (var container = RequestServicesContainer.EnsureRequestServices(httpContext, _services))
+            // All done if we already have a request services
+            if (httpContext.RequestServices != null)
             {
-                await _next.Invoke(httpContext);
+                return;
+            }
+
+            // Resolve the ScopeFactory from the correct SP
+            var priorApplicationServices = httpContext.ApplicationServices;
+            var serviceProvider = priorApplicationServices ?? _services;
+            var appServiceProvider = serviceProvider.GetRequiredService<IServiceProvider>();
```

Comment:
Why not something like:

```
var appServiceProvider  = httpContext.ApplicationServices ?? _services;
```

Why do we call GetService on the IServiceProvider? Doesn't it just return itself?